apps/sms/views/conversation/js/conversation.js

```diff
@@ -1713,14 +1715,24 @@ var ConversationView = {
       pElement.setAttribute('data-l10n-id', 'no-attachment-text');
     }
 
+    this.buildingMessages[message.id] = messageDOM;
+
+    var messageReadyPromise;
     if (message.type === 'mms' && !isNotDownloaded && !noAttachment) { // MMS
-      return this.mmsContentParser(message).then((mmsContent) => {
-        pElement.appendChild(mmsContent);
-        return messageDOM;
-      });
+      messageReadyPromise = this.mmsContentParser(message).then(
+        (mmsContent) => pElement.appendChild(mmsContent)
+      );
+    } else {
+      messageReadyPromise = Promise.resolve();
     }
 
-    return Promise.resolve(messageDOM);
+    return messageReadyPromise.catch(
+      (e) => console.error('Error while building the message DOM', e)
+    ).then(
+      () => delete this.buildingMessages[message.id]
+    ).then(
+      () => messageDOM
```

Comment:
BTW I was thinking do we need to return DOM even the mms parsing is failed. The original implementation didn't return it, but returning the DOM here looks better because we can still return the incomplete message DOM with error message.